File: python/examples/core/06-llm-integration/responder.py
"""Example 06 responder: tiny diagnostic for the Ollama side of the pair.

Polls ``OLLAMA_URL/api/tags`` and prints the available models. Useful for
verifying that Ollama is reachable and that the model the watcher will ask
for is actually installed.
"""

from __future__ import annotations

import asyncio
import os
import sys

import httpx
import structlog

from canvus_sdk import configure_logging

logger = structlog.get_logger(__name__)


def _ollama_url() -> str:
    return os.environ.get("OLLAMA_URL", "http://localhost:11434").rstrip("/")


async def main() -> int:
    """Entry point. Returns 0 if Ollama responded, 1 otherwise."""
    configure_logging()
    url = f"{_ollama_url()}/api/tags"
    logger.info("polling ollama", url=url)
    try:
        async with httpx.AsyncClient(timeout=10.0) as http:
            resp = await http.get(url)
            resp.raise_for_status()
    except httpx.HTTPError as e:
        logger.error("ollama unreachable", url=url, error=str(e))
        return 1

    payload = resp.json()
    models = payload.get("models", [])
    if not models:
        print("(ollama is reachable but no models are installed)")
        return 0

    print(f"models available at {_ollama_url()}:")
    for model in models:
        if not isinstance(model, dict):
            continue
        name = model.get("name", "(unknown)")
        size = model.get("size", 0)
        try:
            size_gb = float(size) / (1024**3)
            size_str = f"{size_gb:.2f} GiB"
        except (TypeError, ValueError):
            size_str = "?"
        modified = model.get("modified_at", "")
        print(f"  {name:<32} {size_str:>12}   {modified}")
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: typescript/examples/core/06-llm-integration/src/responder.ts
/**
 * Example 06 — responder.
 *
 * Lightweight diagnostic for Ollama: lists available models via
 * `/api/tags`. Use this before running the watcher to confirm Ollama
 * is reachable and that the model you want is pulled.
 */

import pino from "pino";
import { z } from "zod";

const isPretty = process.env["LOG_FORMAT"] === "pretty";

const logger = pino({
  level: process.env["LOG_LEVEL"] ?? "info",
  base: { component: "example-llm-responder" },
  ...(isPretty
    ? {
        transport: {
          target: "pino-pretty",
          options: { colorize: true, translateTime: "SYS:HH:MM:ss.l" },
        },
      }
    : {}),
});

const envSchema = z.object({
  OLLAMA_URL: z.string().url().default("http://localhost:11434"),
});

interface OllamaTagsResponse {
  readonly models: readonly {
    readonly name: string;
    readonly size?: number;
    readonly modified_at?: string;
  }[];
}

async function run(): Promise<void> {
  const env = envSchema.parse(process.env);
  const url = new URL("/api/tags", env.OLLAMA_URL);

  logger.info({ url: url.toString() }, "querying Ollama tags");
  const res = await fetch(url);
  if (!res.ok) {
    const text = await res.text();
    logger.error(
      { status: res.status, body: text },
      "Ollama returned non-2xx — is the server running?",
    );
    process.exit(1);
  }
  const payload = (await res.json()) as OllamaTagsResponse;
  if (payload.models.length === 0) {
    logger.warn(
      { url: url.toString() },
      "Ollama is reachable but has no models — run `ollama pull <model>` first",
    );
    return;
  }
  logger.info({ modelCount: payload.models.length }, "available models");
  for (const model of payload.models) {
    process.stdout.write(`  - ${model.name}\n`);
  }
}

run().catch((err: unknown) => {
  logger.error({ err }, "responder failed");
  process.exit(1);
});
